https://issues.redhat.com/browse/ACM-14417 Ver. Strategy doc (#7222)
open-cluster-management/rhacm-docs: clusters/cluster_lifecycle/adv_config_cluster.adoc

  hubKubeAPIServerURL: "example.global.com"
----

[#config-hub-kube-api-server]
== Configuring the hub cluster `KubeAPIServer` verification strategy

Managed clusters communicate with the hub cluster through a mutual connection with the {ocp-short} `KubeAPIServer` external load balancer. An internal {ocp-short} cluster certificate authority (CA) issues the default {ocp-short} `KubeAPIServer` certificate when you install {ocp-short}. The {mce} automatically detects and adds the certificate to managed clusters in the `bootstrap-kubeconfig-secret` namespace.

> If your automatically detected certificate does not work, you can manually configure a strategy configuration in the `KlusterletConfig` resource. Manually configuring the a strategy allows you to control how you verify the hub cluster `KubeAPIServer` certificate.

Reviewer: ```suggestion
If your automatically detected certificate does not work, you can manually configure a strategy configuration in the `KlusterletConfig` resource. Manually configuring the strategy allows you to control how you verify the hub cluster `KubeAPIServer` certificate.
```

See the examples in one of the following three strategies to learn how to manually configure a strategy:

> [#config-hub-kube-api-server-detect]

Reviewer 2: I would update this anchor
```suggestion
[#config-hub-strategy-auto-detect-ca]
```

> See the examples in one of the following three strategies to learn how to manually configure a strategy:
> 
> [#config-hub-strategy-auto-detect-ca]
> === Configuring the strategy with `UseAutoDetectedCABundle`

Reviewer 2: Here's an example of what I was talking about. Related comment: https://github.com/stolostron/rhacm-docs/pull/7222/files#r1844139775

The tricky part is adding callouts to describe the values in the YAML sample. We can probably still keep the examples, and the only line reduction would come from removing the anchors and titles

```suggestion
See the examples in one of the following three strategies to learn how to manually configure a strategy:

`UseAutoDetectedCABundle`:: The default configuration strategy is `UseAutoDetectedCABundle`. The {mce-short} automatically detects the certificate on the hub cluster and merges the certificate configured in the `trustedCABundles` list of config map references to the real CA bundles, if there are any.

`UseSystemTruststore`:: With `UseSystemTruststore`, {mce-short} does not detect any certificate and ignores the certificates configured in the `trustedCABundles` parameter section. This configuration does not pass any certificate to the managed clusters. Instead, the managed clusters use certificates from the system trusted store of the managed clusters to verify the hub cluster API server. This applies to situations where a public CA, such as `Let's Encrypt`, issues the hub cluster certificate.

`UseCustomCABundles`:: You can use `UseCustomCABundles` if you know the CA of the hub cluster API server and do not want {mce-short} to automatically detect it. For this strategy, {mce-short} adds your configured certificates from the `trustedCABundles` parameter to the managed clusters.
```

Author: Thanks for sharing an example. As you mentioned, integrating the YAML samples will be difficult. I'm almost leaning more to keeping the titles, since those make it a bit clearer that there are steps that the user needs to do. Since there are only 3 titles , are you OK with just keeping it as is?

> `UseAutoDetectedCABundle` is the default strategy. The {mce-short} automatically detects the certificate on the hub cluster and merges the certificate configured in the `trustedCABundles` list of ConfigMap references to the real CA bundles, if there are any.

Reviewer 2: Let's try to not start the sentence with the parameter value. Since there are a few back-to-back statements that start with "the", I can understand why you might keep it as is. 
```suggestion
The default configuration strategy is `UseAutoDetectedCABundle`. The {mce-short} automatically detects the certificate on the hub cluster and merges the certificate configured in the `trustedCABundles` list of config map references to the real CA bundles, if there are any.
```

> The following example merges the automatically detected certificates from the hub cluster, and the certificates that you configured in the `new-ocp-ca` ConfigMap, and adds them to the managed cluster:

Reviewer 2: Are both the automatically detected certificates and configured certificates merged?

```suggestion
The following example merges the automatically detected certificates from the hub cluster and the certificates that you configured in the `new-ocp-ca` config map, and adds them to the managed cluster:
```

Author: @zhujian7 Can you help clarify? I'm not sure if both are merged

Reviewer: yes, both are merged, and then pass them to the managed cluster.

[source,yaml]
----
apiVersion: config.open-cluster-management.io/v1alpha1
kind: KlusterletConfig
metadata:
 name: ca-strategy
spec:
 hubKubeAPIServerConfig:
   serverVerificationStrategy: UseAutoDetectedCABundle
   trustedCABundles:
   - name: new-ca
     caBundle:
       name: new-ocp-ca
       namespace: default
----

> [#config-hub-kube-api-server-system]

Reviewer 2: ```suggestion
[#config-hub-strategy-system-truststore]
```

=== Configuring the strategy with `UseSystemTruststore`

> With `UseSystemTruststore`, {mce-short} does not detect any certificate and ignores the certificates configured in the `trustedCABundles`. This configuration does not pass any certificate to the managed clusters. Instead, the managed clusters use certificates in their system trusted store to verify the hub cluster API server. This applies to situations where a public CA, such as `Let's Encrypt`, issues the hub cluster certificate. See the following example that uses `UseSystemTruststore`:

Reviewer 2: Suggestion based on tone and removing anthropomorphism. From my perspective, using "their" is an example of using anthropomorphism. But let me know your perspective. Is "Let's Encrypt" a parameter value? When I googled, I saw that it was a CA. Were you trying to add emphasis to the name? 

```suggestion
With `UseSystemTruststore`, {mce-short} does not detect any certificate and ignores the certificates configured in the `trustedCABundles` parameter section. This configuration does not pass any certificate to the managed clusters. Instead, the managed clusters use certificates from the system trusted store of the managed clusters to verify the hub cluster API server. This applies to situations where a public CA, such as `Let's Encrypt`, issues the hub cluster certificate. See the following example that uses `UseSystemTruststore`:
```

Author: I agree, we can just say `the` instead of `their`.

Let's Encrypt is a name as you already found out. To prevent translation issues I wanted it to be monospaced. Let me know if that makes sense, or if there is a better way to do this.

[source,yaml]
----
apiVersion: config.open-cluster-management.io/v1alpha1
kind: KlusterletConfig
metadata:
 name: ca-strategy
spec:
 hubKubeAPIServerConfig:
   serverVerificationStrategy: UseSystemTruststore
----

> [#config-hub-kube-api-server-custom]

Reviewer 2: ```suggestion
[#config-hub-strategy-custom-ca]
```

=== Configuring the strategy with `UseCustomCABundles`

> You can use `UseCustomCABundles` if you know the CA of the hub cluster API server and do not want {mce-short} to automatically detect it. For this strategy, {mce-short} adds your configured certificates from `trustedCABundles` to the managed clusters. See the following examples to learn how to use `UseCustomCABundles`:

Reviewer 2: ```suggestion
You can use `UseCustomCABundles` if you know the CA of the hub cluster API server and do not want {mce-short} to automatically detect it. For this strategy, {mce-short} adds your configured certificates from the `trustedCABundles` parameter to the managed clusters. See the following examples to learn how to use `UseCustomCABundles`:
```

> You can use `UseCustomCABundles` if you lknow the CA of the hub cluster API server and do not want {mce-short} to automatically detect it. For this strategy, {mce-short} adds your configured certificates from `trustedCABundles` to the managed clusters. See the following examples to learn how to use `UseCustomCABundles`:

Reviewer: ```suggestion
You can use `UseCustomCABundles` if you know the CA of the hub cluster API server and do not want {mce-short} to automatically detect it. For this strategy, {mce-short} adds your configured certificates from `trustedCABundles` to the managed clusters. See the following examples to learn how to use `UseCustomCABundles`:
```

[source,yaml]
----
apiVersion: config.open-cluster-management.io/v1alpha1
kind: KlusterletConfig
metadata:
 name: global
spec:
 hubKubeAPIServerConfig:
   serverVerificationStrategy: UseSystemTruststore
----

When a managed cluster needs to use a different strategy, you can also create a different `KlusterletConfig` and use the `agent.open-cluster-management.io/klusterlet-config` annotation in the managed clusters to point to a specific strategy. See the following example:

[source,yaml]
----
apiVersion: config.open-cluster-management.io/v1alpha1
kind: KlusterletConfig
metadata:
 name: test-ca
spec:
 hubKubeAPIServerConfig:
   serverVerificationStrategy: UseCustomCABundles
   trustedCABundles:
   - name: ca
     caBundle:
       name: ocp-ca
       namespace: default
--
apiVersion: cluster.open-cluster-management.io/v1
kind: ManagedCluster
metadata:
  annotations:
    agent.open-cluster-management.io/klusterlet-config: test-ca
  name: cluster1
spec:
  hubAcceptsClient: true
  leaseDurationSeconds: 60
----

> 

Reviewer 2: Some of the examples are pretty similar with the exception of a different `serverVerificationStrategy` value. What do you think about creating a definition list for the options instead of having a new sub section for each one? Just wanted to know your thoughts. I figured it might reduce some lines, but maybe not lol

Author: Do you have an example? I agree that it's a little long. I wanted to do bullet points at first but remembered that we can't use those for headings due to accessibility. Maybe a definition list is what I was thinking of needing. If you have a good example in mind I'll try and see if I can restructure this section.
